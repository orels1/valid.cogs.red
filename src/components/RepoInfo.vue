<script>
import Vue from 'vue';
import Component from 'vue-class-component';
import { omit } from 'lodash';

@Component({
  props: {
    code: { type: Object, default: () => ({}) },
    type: { type: String, default: 'default' },
    status: { type: String },
  },
})
export default class RepoInfo extends Vue {
  get barStyle() {
    return this.$style[this.type];
  }

  render() {
    const code = omit(this.code, ['DISABLED']);
    const info = Object.keys(code).map((key) => {
      if (!code[key]) return null;
      return (
        <div>
          <b>{key}: </b>{Array.isArray(code[key]) ? code[key].join(', ') : code[key]}
        </div>
      );
    });
    return (
      <div class={[this.$style.repoInfo, this.barStyle]}>
        { this.status &&
          <div class={this.$style.statusLine}>{this.status}</div>
        }
        { Object.keys(this.code).length !== 0 &&
          <div class={this.$style.innerBlock}>
            {info}
          </div>
        }
      </div>
    );
  }
}
</script>

<style lang="sass" module>
$gray: #7f8c8d
$lighter: #353740

.repoInfo
  text-align: left
  background: $lighter
  border-radius: 2px
  margin: 10px auto
  width: auto
  max-width: 500px
  display: flex
  font-size: 12px
  flex-direction: column
  box-shadow: 0 0 8px rgba(0,0,0,.1)

  &.default
    border-left: 2px solid $gray
  
  &.success
    border-left: 2px solid #2ecc71
    background: #27ae60
    color: #fff

  &.error
    border-left: 2px solid #e74c3c
    background: #c0392b
    color: #fff

  &.warning
    border-left: 2px solid #f1c40f
    background: #8e750f
    color: #fff

.statusLine
  padding: 10px 15px
  background: rgba(255,255,255,.2)

.innerBlock
  padding: 15px
  line-height: 16px

  b
    opacity: .6
</style>