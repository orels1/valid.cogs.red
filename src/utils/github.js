import { filter, map, compact } from 'lodash';

// const TOKEN = '';
const API_ROOT = 'https://api.github.com';
const headers = {
  // Authorization: `bearer ${TOKEN}`,
  'Content-Type': 'application/json',
};

export const repoInfo = async (username, repo) => {
  const url = `${API_ROOT}/repos/${username}/${repo}/contents/info.json`;
  const response = await fetch(url, { headers });
  if (response.status === 404) return 'notFound';
  const json = await response.json();
  let info = {};
  try {
    info = JSON.parse(atob(json.content));
  } catch (e) {
    return null;
  }
  return info;
};

export const listCogs = async (username, repo) => {
  const response = await fetch(`${API_ROOT}/repos/${username}/${repo}/contents`, { headers });
  if (response.status === 404) return Promise.reject('NotFound');
  const json = await response.json();
  const cogs = filter(json, { type: 'dir' });
  return map(cogs, cog => cog.name);
};

export const cogInfo = async (username, repo, cog, tree = 'master') => {
  const response = await fetch(`${API_ROOT}/repos/${username}/${repo}/contents/${cog}/info.json?ref=${tree}`, { headers });
  if (response.status === 404) return false;
  const json = await response.json();
  let info = {};
  try {
    info = JSON.parse(atob(json.content));
  } catch (e) {
    return null;
  }
  return info;
};

export const parseCogs = async (username, repo, branch = 'master') => {
  let cogs = await listCogs(username, repo);

  const cogsInfo = await Promise.all(cogs.map(c => cogInfo(username, repo, c, branch)));

  cogs = cogs.map((c, index) => ({
    name: c,
    info: cogsInfo[index],
  }));

  const broken = [];
  const missing = [];

  cogs = compact(map(cogs, (c) => {
    if (c.info && c.info !== null) return c;
    if (c.info === null) {
      broken.push(c.name);
      return null;
    }
    missing.push(c.name);
    return null;
  }));

  return { cogs, broken, missing };
};
