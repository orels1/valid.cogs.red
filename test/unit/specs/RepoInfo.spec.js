import Vue from 'vue';
import RepoInfo from '@/components/RepoInfo';
import { render, CssModuleTestHelperMixin } from '../utils';

Vue.use(CssModuleTestHelperMixin);

describe('RepoInfo component', () => {
  it('computes default info property', () => {
    const { wrapper } = render(RepoInfo);
    expect(wrapper.vm.info).toEqual([]);
  });

  it('omits DISABLED field', () => {
    const { wrapper } = render(RepoInfo, { code: {
      DISABLED: true,
      foo: 'bar',
    } });
    expect(wrapper.vm.info).toEqual([{ key: 'foo', val: 'bar' }]);
  });

  it('omits undefined or null fields', () => {
    const { wrapper } = render(RepoInfo, { code: {
      foo: null,
      trigger: undefined,
      fizz: 'buzz',
    } });
    expect(wrapper.vm.info).toEqual([{ key: 'fizz', val: 'buzz' }]);
  });

  it('renders arrays as lists', () => {
    const { wrapper } = render(RepoInfo, { code: {
      TAGS: ['some', 'other', 'tag'],
    } });
    expect(wrapper.vm.info).toEqual([{ key: 'TAGS', val: 'some, other, tag' }]);
  });

  test('sets default props', () => {
    const propsList = ['code', 'type']; // list of default props
    const { wrapper } = render(RepoInfo);
    const props = {};

    propsList.forEach((prop) => {
      props[prop] = wrapper.vm[prop];
    });
    expect(props).toMatchSnapshot();
  });

  describe('status line', () => {
    test('hides by default', () => {
      const { wrapper } = render(RepoInfo);
      expect(wrapper.contains('.stausLine')).toBe(false);
    });

    test('shows if status was provided', () => {
      const { wrapper, props } = render(RepoInfo, {
        status: 'foobar',
      });
      expect(wrapper.contains('.statusLine')).toBe(true);
      expect(wrapper.find('.statusLine').text()).toBe(props.status);
    });
  });

  describe('inner block', () => {
    test('hides by default', () => {
      const { wrapper } = render(RepoInfo);
      expect(wrapper.contains('.innerBlock')).toBe(false);
    });

    test('shows if proper code was supplied', () => {
      const { wrapper } = render(RepoInfo, {
        code: { AUTHOR: 'foobar' },
      });
      expect(wrapper.contains('.innerBlock')).toBe(true);
    });

    test('renders supplied code info', () => {
      const { wrapper } = render(RepoInfo, {
        code: {
          AUTHOR: 'foobar',
          NAME: 'fizzbuzz',
        },
      });
      expect(wrapper.find('.innerBlock').html()).toMatchSnapshot();
    });
  });
});
