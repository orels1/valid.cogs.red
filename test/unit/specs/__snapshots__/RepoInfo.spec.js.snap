// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoInfo component inner block renders supplied code info 1`] = `
<div class="innerBlock">
    <div>
        <b>AUTHOR:</b> foobar</div>
    <div>
        <b>NAME:</b> fizzbuzz</div>
</div>
`;

exports[`RepoInfo component sets default props 1`] = `
Object {
  "code": Object {},
  "type": "default",
}
`;
